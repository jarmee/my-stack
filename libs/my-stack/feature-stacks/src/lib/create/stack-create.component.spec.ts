import { provideHttpClient } from '@angular/common/http';
import { importProvidersFrom } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { faker } from '@faker-js/faker';
import {
  ApiModule,
  CreateStackDto,
  StacksService,
} from '@my-stack/shared/api-my-stack';
import { of } from 'rxjs';

import { StackCreateComponent } from './stack-create.component';

const CSS_SELECTORS = {
  TITLE_FORM_FIELD: By.css('[data-test-id="title-form-field"]'),
  FORM: By.css('form'),
};

function setTitleFormFieldFactory(
  fixture: ComponentFixture<StackCreateComponent>
) {
  return (value: string) => {
    fixture.componentInstance.formGroup.get('title')?.setValue(value);
    fixture.detectChanges();
  };
}

function submitFormFactory(fixture: ComponentFixture<StackCreateComponent>) {
  const submitFormElement = fixture.debugElement.query(CSS_SELECTORS.FORM);

  return () => {
    submitFormElement.triggerEventHandler('ngSubmit', null);
    fixture.detectChanges();
  };
}

describe('StackCreateComponent', () => {
  let component: StackCreateComponent;
  let fixture: ComponentFixture<StackCreateComponent>;
  let service: StacksService;
  let router: Router;
  let setTitleFormField: (value: string) => void;
  let submitForm: () => void;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        StackCreateComponent,
        RouterTestingModule,
        NoopAnimationsModule,
      ],
      providers: [provideHttpClient(), importProvidersFrom(ApiModule)],
    }).compileComponents();

    service = TestBed.inject(StacksService);
    router = TestBed.inject(Router);

    fixture = TestBed.createComponent(StackCreateComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    setTitleFormField = setTitleFormFieldFactory(fixture);
    submitForm = submitFormFactory(fixture);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('form', () => {
    beforeEach(() => {
      service.createStack = jest.fn().mockReturnValue(of(null));
      router.navigate = jest.fn();
    });

    it('should do nothing when the user clicks the submit button and the form is invalid', () => {
      submitForm();

      expect(service.createStack).not.toHaveBeenCalled();
    });
    it('should create a new stack when the user clicks the submit button and the form is valid', () => {
      const expected: CreateStackDto = {
        title: faker.random.words(),
      };

      setTitleFormField(expected.title);
      submitForm();

      expect(service.createStack).toHaveBeenCalledWith(expected);
    });
  });
});
