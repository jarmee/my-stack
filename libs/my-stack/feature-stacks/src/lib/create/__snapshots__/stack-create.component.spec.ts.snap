// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackCreateComponent should match the snapshot 1`] = `
<ng-component
  __ngContext__={[Function Number]}
  _formBuilder={[Function FormBuilder]}
  _router={[Function Router]}
  _service={[Function StacksService]}
  formGroup={[Function FormGroup]}
>
  <div
    class="px-4"
  >
    <div
      class="flex items-end"
    >
      <button
        class="mb-3 mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
        mat-button=""
        routerlink="/"
        tabindex="0"
      >
        <span
          class="mat-mdc-button-persistent-ripple mdc-button__ripple"
        />
        <mat-icon
          aria-hidden="true"
          class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
          data-mat-icon-type="font"
          role="img"
        >
          arrow_back
        </mat-icon>
        <span
          class="mdc-button__label"
        />
        <span
          class="mat-mdc-focus-indicator"
        />
        <span
          class="mat-ripple mat-mdc-button-ripple"
          matripple=""
        />
        <span
          class="mat-mdc-button-touch-target"
        />
      </button>
      <h3>
        Create Stack
      </h3>
    </div>
    <mat-divider
      aria-orientation="horizontal"
      class="mat-divider mat-divider-horizontal"
      role="separator"
    />
    <div
      class="flex justify-center"
    >
      <form
        class="flex flex-col container ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        <mat-form-field
          appearance="fill"
          class="mat-mdc-form-field mt-8 mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-form-field-hide-placeholder mat-primary ng-untouched ng-pristine ng-invalid"
        >
          <div
            class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled"
          >
            <div
              class="mat-mdc-form-field-focus-overlay"
            />
            <div
              class="mat-mdc-form-field-flex"
            >
              <div
                class="mat-mdc-form-field-infix"
              >
                <label
                  class="mdc-floating-label mat-mdc-floating-label"
                  matformfieldfloatinglabel=""
                >
                  <mat-label>
                    Title
                  </mat-label>
                  <span
                    aria-hidden="true"
                    class="mat-mdc-form-field-required-marker mdc-floating-label--required"
                  />
                </label>
                <input
                  aria-required="true"
                  class="mat-mdc-input-element ng-untouched ng-pristine ng-invalid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
                  data-test-id="title-form-field"
                  formcontrolname="title"
                  matinput=""
                  required=""
                />
              </div>
            </div>
            <div
              class="mdc-line-ripple mdc-line-ripple--deactivating"
              matformfieldlineripple=""
            />
          </div>
          <div
            class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
          >
            <div
              class="mat-mdc-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              <div
                class="mat-mdc-form-field-hint-spacer"
              />
            </div>
          </div>
        </mat-form-field>
        <button
          class="mt-3 mdc-button mdc-button--raised mat-mdc-raised-button mat-primary _mat-animation-noopable mat-mdc-button-base"
          color="primary"
          data-test-id="submit-button"
          disabled="true"
          mat-raised-button=""
          type="submit"
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-button__ripple"
          />
          <span
            class="mdc-button__label"
          >
             Create 
          </span>
          <span
            class="mat-mdc-focus-indicator"
          />
          <span
            class="mat-ripple mat-mdc-button-ripple"
            matripple=""
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </button>
      </form>
    </div>
  </div>
</ng-component>
`;
